### PR TITLE
feat(linter): Add po-linter script and dockerfile

### PR DESCRIPTION
**Pull request by jackstockley89**
Adding script that will scan PR files for PrometheusRules and check if command is valid

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,68 @@
-# cloud-platform-po-linter
\ No newline at end of file
+# cloud-platform-po-linter
+### Description
+---
+This linter will check Prometheus Rules YAML for any incorrect syntax.
```
Comment (jasonBirchall) [inline README.md]: I think here we can describe it as a GitHub Action?

File: cloud-platform-po-linter.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	"github.com/ministryofjustice/cloud-platform-environments/pkg/authenticate"
+	"github.com/prometheus-operator/prometheus-operator/pkg/admission"
+	v1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/prometheus/prometheus/model/rulefmt"
+	githubaction "github.com/sethvargo/go-githubactions"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func prDir(token, githubrepo, githubref string) ([]string, error) {
+	if token == "" {
```
Comment (jasonBirchall) [inline cloud-platform-po-linter.go:25]: While we're being defensive here (which is good) - what happens if we pass empty `gitHubRef` and `repo` strings?

File: cloud-platform-po-linter.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+	//repo user and repo name
+	githubrepoS := strings.Split(githubrepo, "/")
+	owner := githubrepoS[0]
+	repo := githubrepoS[1]
+
+	// get pr owner
+	githubrefS := strings.Split(githubref, "/")
+	branch := githubrefS[2]
+	bid, _ := strconv.Atoi(branch)
```
Comment (jasonBirchall) [inline cloud-platform-po-linter.go]: handle the error here.

File: cloud-platform-po-linter.go (hunk 2 above)
Comment (jasonBirchall) [reply cloud-platform-po-linter.go:25]: Also, can I do anon api calls?

File: cloud-platform-po-linter.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+			}
+			err = validateRules(content)
+			if err != nil {
+				// log.Fatalf("prometheus rule validation failed: %v", err)
+				o := fmt.Sprintf("prometheus rule validation failed: %v", err)
+				githubaction.SetOutput("po-linter", o)
+				os.Exit(1)
+			}
+		default:
+			// log.Print("MetaType is unknown to linter. Not in PrometheusRule")
```
Comment (jasonBirchall) [inline cloud-platform-po-linter.go]: Remove commented-out code.

File: cloud-platform-po-linter.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+			o := fmt.Sprintln("MetaType is unknown to linter. Not in PrometheusRule")
+			githubaction.SetOutput("po-linter", o)
```
Comment (jasonBirchall) [inline cloud-platform-po-linter.go]: Should we return the GitHub Action output type? Then we can test.

File: cloud-platform-po-linter.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+		content, err := ioutil.ReadFile(filename)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		var meta metav1.TypeMeta
+
+		err = yaml.Unmarshal(content, &meta)
+		if err != nil {
+			log.Fatal(err)
```
Comment (jasonBirchall) [inline cloud-platform-po-linter.go]: We could return an error here, and let the caller handle it

File: cloud-platform-po-linter.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+	return fileDir, nil
+
+}
+
+func poLint(dir []string) {
+	files := dir
+
+	for _, filename := range files {
+		log.SetPrefix(fmt.Sprintf("%s: ", filename))
+		content, err := ioutil.ReadFile(filename)
```
Comment (jasonBirchall) [inline cloud-platform-po-linter.go]: ioutil was deprecated in 1.16
https://pkg.go.dev/io/ioutil

I think you can use the os package.

File: go.mod [hunk 8]
```diff
@@ -0,0 +1,123 @@
+module github.com/ministryofjustice/cloud-platform-po-linter
+
+go 1.18
```
Comment (jasonBirchall) [inline go.mod]: Bump this to 1.20?

Comment (jackstockley89) [reply go.mod]: do you mean 1.19?

Comment (jasonBirchall) [reply go.mod]: Yep, sorry 1.19

**Review by jasonBirchall — APPROVED** [review]

**Pull request merged**